Add Lua Scripting for Carryall.

MustaphaTR opened the pull request:
I changed the Carryalls in SP to actually use Carryall logic. The shellmap had some carryall stuff, which we didn't support yet for the actual Carryall logic. This PR adds functions for Picking Up and Dropping the carried unit for Carryalls. Altho i'm not sure if what i did for ignoring the Pickup/Deliver if the Carryall is already Full/Empty is the best way of doing it.

MustaphaTR committed:
Add Lua Scripting for Carryall.

pchote commented:
There are a couple of issues here, demonstrated with a testcase:

* Add the following to `sunstroke.lua`:
   ```lua
   WorldLoaded = function()
   	Multi = Player.GetPlayer("Multi0")
   	local carryall = Actor.Create("TRNSPORT", true, { Owner = Multi, Location = Actor615.Location })
   	local cargo = Actor.Create("APC", true, { Owner = Multi, Location = Actor615.Location + CVec.New(3, 3) })
   	carryall.PickupUnit(cargo)
   	carryall.Move(Actor615.Location - CVec.New(3, 3))
   	carryall.DeliverUnit()
   end
   ```
* Start a game at the A spawn
* Select the carryall and hold shift to watch target lines

The first problem is that this this generates target lines, whereas the standard behaviour for scripted activities (whether we like it or not - this isn't the place to argue it) is to hide them by passing `null` to a `targetLineColor` arg in the activity constructors. This will need to be hooked up for `PickupUnit` and `DeliverUnit`.

The second problem is that the `if (carryall.Carryable == null)` check in `DeliverUnit` runs immediately, before the cargo is loaded, so delivery is skipped. This should be deferred until the activity runs using the `CallFunc` activity.

Can we please also call the Lua APIs for these `PickupCarryable` and `DropCarryable`? "Unit" is too general for the actor namespace and "Deliver" implies it should take a location as an argument. This isn't the first time that we hide the true activity names from the API.

abcdefg30 commented:
> This should be deferred until the activity runs using the `CallFunc` activity.

I'm not quite sure what the suggestion here is. Having the checks at all seems unnecessary. (And shouldn't be worked around in Lua code using `CallFunc`?) The activities should handle that case on their own.

MustaphaTR commented:
Updated. Fixed the first and 3rd points as PChote said. I removed the checks as abcdefg30 said, but haven't checked/updated the activities to properly check them. I have some questions first, will handle that along with them.

1. Should AutoCarryall and FreeActorWithDelivery have target lines? It doesn't matter much for base mods, since our only usecase for those traits are not selectable. But also they are automatic orders, that are not done by the player, similar to lua scripts.
2. I've just noticed DeliverUnit has a second constructor that has a target arguement. If i change the lua function to have an optional target arguement and use either depending on if it is defined, should the function still be called DropCarryable or is DeliverCarryable better then. Should the arguement be optional at the first place?

pchote commented:
1. I'm not sure, and I think we risk derailing this PR if we go too far down that topic (I'm personally not convinced about disabling target lines at all)
2. I think it would be best to have a single `DeliverCarryable` that always requires a destination.

MustaphaTR commented:
Updated with the point 2. Activities already seem to be properly ignoring the Pickup/Deliver order if they are already Full/Empty. You can see with this version of the testcase PChote posted above.
```lua
WorldLoaded = function()
	Multi = Player.GetPlayer("Multi0")
	local carryall = Actor.Create("TRNSPORT", true, { Owner = Multi, Location = Actor615.Location })
	local apc = Actor.Create("APC", true, { Owner = Multi, Location = Actor615.Location + CVec.New(3, 3) })
	local bike = Actor.Create("BIKE", true, { Owner = Multi, Location = Actor615.Location + CVec.New(0, 3) })
	carryall.PickupCarryable(apc)
	carryall.PickupCarryable(bike)
	carryall.DeliverCarryable(Actor615.Location - CVec.New(3, 3))
	carryall.DeliverCarryable(Actor615.Location + CVec.New(3, 3))
end
```

MustaphaTR commented:
This has been sitting for a while, i would like to note again that this was updated and the label is inaccurate. I can't edit labels anymore.

abcdefg30 commented:
This has compiler errors once you rebase.

MustaphaTR commented:
Rebased and updated.

penev92 commented on OpenRA.Mods.Common/Traits/AutoCarryall.cs:
```diff
@@ -114,18 +114,19 @@ public FerryUnit(Actor self, Actor cargo)
 			{
 				if (!cargo.IsDead)
-					QueueChild(new PickupUnit(self, cargo, 0));
+					QueueChild(new PickupUnit(self, cargo, 0, self.Trait<Carryall>().Info.TargetLineColor));
```
You should add a `readonly CarryallInfo carryallInfo;` field and do a single lookup in the constructor (because currently there in one on every tick + this initial one).

penev92 reviewed (dismissed):
LGTM with one small fix request

penev92 reviewed (approved):
👍

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9049d0937d177dd9c54566cb3aef5a553062e2da...7a32c46bedfc8448db1dfc37c705a67748552075)
